File: ports/csharp/src/gui/widgets/BcsLayout.cs
using System.Collections.Generic;
using Bcs.Core.Kernel;

namespace Bcs.Gui.Widgets
{
    public class BcsLayout : BcsObject
    {
        private List<BcsWidget> _widgets;
        private readonly object _lock = new object();

        public BcsLayout(BcsObject parent = null) : base(parent)
        {
            _widgets = new List<BcsWidget>();
        }

        public void AddWidget(BcsWidget widget)
        {
            lock (_lock)
            {
                _widgets.Add(widget);
            }
        }

        public void RemoveWidget(BcsWidget widget)
        {
            lock (_lock)
            {
                _widgets.Remove(widget);
            }
        }

        public List<BcsWidget> Widgets()
        {
            lock (_lock)
            {
                return new List<BcsWidget>(_widgets);
            }
        }
    }
}

File: ports/rust/src/gui/widgets/bcs_layout.rs
use std::sync::{Arc, Mutex};
use crate::core::kernel::bcs_object::BcsObject;
use crate::gui::widgets::bcs_widget::BcsWidget;

pub struct BcsLayout {
    pub base: Arc<BcsObject>,
    widgets: Mutex<Vec<Arc<BcsWidget>>>,
}

impl BcsLayout {
    pub fn new(parent: Option<Arc<BcsObject>>) -> Arc<Self> {
        let base = if let Some(p) = parent {
            BcsObject::new_with_parent(p)
        } else {
            BcsObject::new()
        };

        Arc::new(Self {
            base,
            widgets: Mutex::new(Vec::new()),
        })
    }

    pub fn add_widget(&self, widget: Arc<BcsWidget>) {
        let mut widgets = self.widgets.lock().unwrap();
        widgets.push(widget);
    }

    pub fn remove_widget(&self, widget: &Arc<BcsWidget>) {
        let mut widgets = self.widgets.lock().unwrap();
        if let Some(pos) = widgets.iter().position(|x| Arc::ptr_eq(x, widget)) {
            widgets.remove(pos);
        }
    }

    pub fn widgets(&self) -> Vec<Arc<BcsWidget>> {
        let widgets = self.widgets.lock().unwrap();
        widgets.clone()
    }
}
